### PR TITLE
Elevate erg

### PR DESCRIPTION
**Pull request by erg**
Test me, baby!

Rebased and fixed circularity (?) on mac...

Same as one patch more + https://github.com/factor/factor/pull/2000

### CODE REVIEW

File: basis/elevate/macosx/macosx.factor [hunk 1]
```diff
@@ -0,0 +1,28 @@
+USING: accessors arrays cocoa.apple-script elevate
+elevate.unix.private formatting io.launcher kernel locals
+sequences system ;
+IN: elevate.macosx
+
+<PRIVATE
+: apple-script-elevated ( command -- )
+    first quote-apple-script
+    "do shell script %s with administrator privileges without altering line endings"
```
Comment (erg) [inline basis/elevate/macosx/macosx.factor]: The ``%s`` here needs to be rethought. We have a command as ``{ "ls" "-al" }`` and that doesn't work with the format string here.

File: basis/elevate/macosx/macosx.factor [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+! TODO
+M:: macosx elevated ( command replace? win-console? posix-graphical? -- process )
+    already-root? [
+        <process> command >>command 1array
+    ] [
+        ! graphical through applescript
+        posix-graphical? [
+            command apple-script-elevated
+        ] when
+        posix-elevated  "lol3" throw
```
Comment (erg) [inline basis/elevate/macosx/macosx.factor]: i left in

File: basis/elevate/macosx/macosx.factor [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    ] if "lol" throw ;
```
Comment (erg) [inline basis/elevate/macosx/macosx.factor]: some debugging...oops

File: basis/elevate/macosx/macosx.factor (hunk 1 above)
Comment (catb0t) [reply basis/elevate/macosx/macosx.factor]: https://github.com/barneygale/elevate/blob/master/elevate/posix.py#L37

Comment (catb0t) [reply basis/elevate/macosx/macosx.factor]: the format argument in the Python is `quote_applescript(quote_shell(args))`, which is supposed to be replaced by `quote-apple-script` but I couldn't find our `quote-shell` word even though i know we have one

**Comment by catb0t** [comment]
GKSudo is deprecated, and needs to be demoted from the primary method on Linux. PKExec is very complicated to get working properly and more code is needed (https://wiki.archlinux.org/index.php/Running_GUI_apps_as_root)

**Comment by catb0t** [comment]
for GUI applications to work with PKExec there needs to be an .xml file in a system folder, so we have to do a "first-time setup" thing where we ask to create the file  https://askubuntu.com/questions/288087/can-i-use-pkexec-in-a-python-script-or-a-desktop-file

**Comment by catb0t** [comment]
I don't understand why the CI is failing for this also? None of the tests failed and none of the programs exited with 1.

 I'm gonna pull this branch but I think maybe if you push again the CI might run properly?

**Comment by mrjbq7** [comment]
Merged, thanks.
